restcomm/restcomm.interpreter/src/main/java/org/restcomm/connect/interpreter/NumberSelectorService.java

```diff
@@ -0,0 +1,245 @@
+            if (logger.isDebugEnabled()) {
+                logger.debug("Matched number with filter:" + matchedNumbers.get(0));
+            }
+            matchedNumber = matchedNumbers.get(0);
+        }
+        return matchedNumber;
+    }
+
+    private IncomingPhoneNumber findByNumber(List<String> numberQueries,
+            Sid sourceOrganizationSid, Sid destinationOrganizationSid) {
+        IncomingPhoneNumber matchedNumber = null;
+
+        int i = 0;
+        while (matchedNumber == null && i < numberQueries.size()) {
+            matchedNumber = findSingleNumber(numberQueries.get(i),
+                    sourceOrganizationSid, destinationOrganizationSid);
+            i = i + 1;
+        }
+        return matchedNumber;
+    }
+
+    public IncomingPhoneNumber searchNumber(String phone,
+            Sid sourceOrganizationSid, Sid destinationOrganizationSid) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("getMostOptimalIncomingPhoneNumber: " + phone
+                    + ",srcOrg:" + sourceOrganizationSid
+                    + ",destOrg:" + destinationOrganizationSid);
+        }
+        List<String> numberQueries = createPhoneQuery(phone);
+
+        IncomingPhoneNumber numberfound = findByNumber(numberQueries, sourceOrganizationSid, destinationOrganizationSid);
+        if (numberfound == null) {
+            //if no perfect match found, try with special star number
+            numberfound = findSingleNumber("*", sourceOrganizationSid, destinationOrganizationSid);
+            if (numberfound == null) {
+                //only use regex if perfect match didnt worked
+                if (destinationOrganizationSid != null
+                        && phone.matches("[\\d,*,#,+]+")) {
+                    //only use regex if org available
+                    //check if there is a Regex match only if parameter is a String aka phone Number
+                    numberfound = findByRegex(numberQueries, sourceOrganizationSid, destinationOrganizationSid);
+                }
+            }
+        }
+        return numberfound;
+    }
+
+    class NumberLengthComparator implements Comparator<IncomingPhoneNumber> {
+
+        @Override
+        public int compare(IncomingPhoneNumber o1, IncomingPhoneNumber o2) {
+            //put o2 first to make longest first in coll
+            return Integer.compare(o2.getPhoneNumber().length(), o1.getPhoneNumber().length());
+        }
+
+    }
+
+    /**
+     * @param numbers
+     * @param phone
+     * @param numbersDao
+     * @return
+     */
+    private IncomingPhoneNumber findByRegex(List<String> numberQueries,
+            Sid sourceOrganizationSid, Sid destOrg) {
+        IncomingPhoneNumber numberFound = null;
+        IncomingPhoneNumberFilter.Builder filterBuilder = IncomingPhoneNumberFilter.Builder.builder();
+        filterBuilder.byOrgSid(destOrg.toString());
+        filterBuilder.byPureSIP(Boolean.FALSE);
```

Comment:
@jaimecasero why we exclude pure sip numbers here? 
The Regex numbers will be created as pure sip numbers since they will be created manually and not through a Voip provider integration

Reply:
here we are getting the regexes, that are expected not to be pureSIP. This is to let the rDBMS discard lots of rows to compare, since the way we get regexes is with a very awful combination of Like operators.

Reply:
@jaimecasero I am might be missing something here but the IncomingPhoneNumbers created for Regex rules will be `pureSip` numbers.
The Regex IncomingPhoneNumbers will be created using IncomingPhoneNumber REST API and providing `pureSip=true` or using Dashboard `sip number` tab

Reply:
@gvagenas umm, i expected regexes to be created as pureSIP=false, since I assumed pureSIP=true was for sipURIs numbers... let me check that again.

Reply:
@jaimecasero You are right, my misunderstanding, I mixed up the pureSip flag role

Reply:
@jaimecasero the requirement here is to differentiate provider numbers (Nexmo) from the rest. For provider numbers we need to follow different flow to purchase it from the provider. Other than that all numbers are the same and pureSip flag is enough(before regex) to my opinion to cover this requirement. With the REGEX and the Smpp restrictions, we might need another flag for REGEX numbers

Reply:
changed to pureSIP=true